models/demos/utils/llm_demo_utils.py

```diff
@@ -0,0 +1,123 @@
+    - The N_warmup_iter dictionary should contain the following keys: "inference_prefill", "inference_decode"
+
+    Optional (should be provided if doing token verification, not required for perf):
+    - The measurements and targets dictionaries should contain the key "token_verification"
+    """
+
+    assert all(
+        key in measurements
+        for key in [
+            "compile_prefill",
+            "compile_decode",
+            "prefill_t/s",
+            "decode_t/s",
+            "decode_t/s/u",
+        ]
+    )
+
+    benchmark_data = BenchmarkData()
+
+    # Add required measurement data
+    benchmark_data.add_measurement(profiler, 0, "compile_prefill", "time(s)", measurements["compile_prefill"])
+    benchmark_data.add_measurement(profiler, 0, "compile_decode", "time(s)", measurements["compile_decode"])
+    benchmark_data.add_measurement(
+        profiler,
+        0,
+        "inference_prefill",
+        "tokens/s",
+        measurements["prefill_t/s"],
+        step_warm_up_num_iterations=N_warmup_iter["inference_prefill"]
+        if "inference_prefill" in N_warmup_iter
+        else None,
+        target=targets["prefill_t/s"] if "prefill_t/s" in targets else None,
+    )
+    benchmark_data.add_measurement(
+        profiler,
+        0,
+        "inference_decode",
+        "tokens/s",
+        measurements["decode_t/s"],
+        step_warm_up_num_iterations=N_warmup_iter["inference_decode"] if "inference_decode" in N_warmup_iter else None,
+        target=targets["decode_t/s"] if "decode_t/s" in targets else None,
+    )
+    benchmark_data.add_measurement(
+        profiler,
+        0,
+        "inference_decode",
+        "tokens/s/user",
+        measurements["decode_t/s/u"],
+        step_warm_up_num_iterations=N_warmup_iter["inference_decode"] if "inference_decode" in N_warmup_iter else None,
+        target=targets["decode_t/s/u"] if "decode_t/s/u" in targets else None,
+    )
+
+    # Add optional token verification datapoint
+    if "token_verification" in measurements:
+        benchmark_data.add_measurement(
+            profiler,
+            0,
+            "inference_decode",
+            "token_verification",
+            measurements["token_verification"],
+            target=targets["token_verification"] if "token_verification" in targets else None,
+        )
+
+    return benchmark_data
+
+
+def check_tokens_match(generated_text: dict, expected_greedy_output_path: str):
+    with open(expected_greedy_output_path, "r") as f:
+        expected_output = json.load(f)
+    return generated_text == expected_output, expected_output
+
+
+def verify_perf(measurements: dict, expected_perf_metrics: dict):
+    """
+    Verify the performance metrics against the expected values.
+    The metrics that must be provided are specified in expected_measurements below.
+    """
+
+    expected_measurements = {
```

Comment:
Same here. What if a test does not have prefill or decode? Should the function just verify against what's in the target, rather than enforcing certain metrics to be available?

Reply:
Same comment as above, prefill/decode measurements are required for llm demos.